File: 2020/day3.js
'use strict';

const fs = require('fs');

function main() {
	const lines = fs.readFileSync('input.txt').toString().split('\r\n');

	const map = [];

	for(const line of lines) {
		const row = [];
		for(let i = 0; i < line.length; i++) {
			row.push(line.charAt(i));
		}
		map.push(row);
	}

	// Part 1
	// return travel(map, 3, 1);

	// Part 2
	return travel(map, 1, 1) * travel(map, 3, 1) * travel(map, 5, 1) * travel(map, 7, 1) * travel(map, 1, 2);
}

function travel(map, right, down) {
	let num = 0, height = 0, length = 0;
	const full_len = map[0].length;

	while(height < map.length) {
		if(map[height][length] === '#') {
			num++;
		}

		length += right;
		height += down;

		if(length >= full_len) {
			length = length % full_len;
		}
	}
	return num;
}

console.log(main());

File: 2016/src/bin/day3.rs
use std::fs::File;
use std::io::prelude::*;
use std::path::Path;

fn parse_ints(line: String) -> Vec<i32> {
    line.trim().split_whitespace().map(|p| p.parse::<i32>().unwrap()).collect::<Vec<_>>()
}

fn main() {
    let path = Path::new("input.txt");
    let display = path.display();

    let mut file = match File::open(&path) {
        Err(why) => panic!("couldn't open {}: {}", display, why),
        Ok(file) => file,
    };

    let mut s = String::new();
    file.read_to_string(&mut s).unwrap();

    let lines = s.trim().split("\n");
    let mut valid_triangles = 0;

    for line in lines.clone() {
        let mut parts = parse_ints(line.to_string());
        parts.sort();

        if parts[0] + parts[1] > parts[2] {
            valid_triangles += 1;
        }
    }

    println!("part 1 {}", valid_triangles);

    valid_triangles = 0;

    for line_group in lines.collect::<Vec<_>>().chunks(3) {
        let mut triangles = vec!(Vec::new(), Vec::new(), Vec::new());

        for line in line_group {
            let parts = parse_ints(line.to_string());

            for i in 0..3 {
                triangles[i].push(parts[i]);
            }
        }

        for i in 0..3 {
            triangles[i].sort();

            if triangles[i][0] + triangles[i][1] > triangles[i][2] {
                valid_triangles += 1;
            }
        }
    }

    println!("part 2 {}", valid_triangles);
}